src/traversals/eulerian.jl

```diff
@@ -0,0 +1,93 @@
+# Adapted from SimpleGraphs.jl [Copyright (c) 2014, Ed Scheinerman]:
+#    https://github.com/scheinerman/SimpleGraphs.jl/blob/master/src/simple_euler.jl
+# Reproduced under the MIT Expat License.
+
+"""
+    eulerian(g::AbstractSimpleGraph{T}[, u::T]) --> T[]
+
+Returns a [Eulerian trail or cycle](https://en.wikipedia.org/wiki/Eulerian_path) through an
+undirected graph `g`, starting at `u`, returning a vector listing the vertices of `g` in the
+order that they are traversed. If no such trail or cycle exists, throws an error.
+
+## Optional arguments
+- If `u` is omitted, a Eulerian trail or cycle is computed with `u = first(vertices(g))`.
+"""
+function eulerian(g::AbstractSimpleGraph{T}, u::T=first(vertices(g))) where {T}
+    is_directed(g) && error("`eulerian` is not yet implemented for directed graphs")
+    
+    _check_eulerian_input(g, u) # perform basic sanity checks
+
+    g′ = SimpleGraph{T}(nv(g)) # copy `g`
+    for e in edges(g)
+        add_edge!(g′, e)
+    end
+
+    return _eulerian!(g′, u)
+end
+
+function _eulerian!(g::AbstractSimpleGraph{T}, u::T) where {T}
+    # TODO: This uses Fleury's algorithm which is O(|E|²) in the number of edges |E|.
+    #       Hierholzer's algorithm [https://en.wikipedia.org/wiki/Eulerian_path#Hierholzer's_algorithm]
+    #       is presumably faster, running in O(|E|) time, but requires needing to keep track
+    #       of visited/nonvisited sites in a doubly-linked list/deque.
+    trail = T[]
+
+    nverts = nv(g)
+    while true
+        # if last vertex
+        if nverts == 1
+            push!(trail, u)
+            return trail
+        end
+
+        Nu = neighbors(g, u)
+        if length(Nu) == 1
+            # if only one neighbor, delete and move on
+            w = first(Nu)
+            rem_edge!(g, u, w)
+            nverts -= 1
+            push!(trail, u)
+            u = w
+        else
+            # otherwise, pick whichever neighbor is not a bridge/cut-edge
+            bs = bridges(g)
```

Comment:
Is there a way to compute bridges relative to a vertex?

Reply:
Yeah, the repeated calls to `bridges` is what gives the Fleury algorithm the O(E^2) scaling. This could be avoided with Hierholzer's algorithm, in principle. From what I can tell, there isn't currently a way to compute bridges relative to a vertex in Graphs.jl: would be cool to have though.